### DISCUSSION TITLE
Duplicated codes improvement

### DISCUSSION THREAD

**Discussion by nima70** [keycloakify/keycloakify#708] [Ideas]
Hi @garronej,

I've noticed some duplicated code in the project, particularly with components like the password holder. In my fork, I refactored these to streamline the code, and I was wondering if you'd be open to a similar refactor in Keycloakify.

This might break some other people’s projects, but it would simplify the styling process. Instead of styling the same element repeatedly, we could manage it all in one place.

This is just a suggestion! I’m not really an open-source guy, so I’m not sure if this is a good idea or not. However, in my experience in the corporate world, existing users can sometimes be resistant to these kinds of changes.

**Comment by garronej** [comment]
Hello @nima70,  

You're right; this is indeed a significant challenge for Keycloakify.

The core issue can't be tackled incrementally. The real difficulty isn’t just in breaking the code into smaller components—it’s in ensuring the eject commands work seamlessly.

Assuming we have factorized the pages into smaller components, the eject command implementation can't simply be a script that copies page sources from the Keycloakify module into the user’s codebase anymore. It will need to analyze which components are in use and copy those as well, recursively.  

That alone is complex. However, it still doesn't fully resolve the issue, as non-ejected pages will continue relying on internal components. To address this, we would need a mechanism to pass component implementations to pages, similar to how we handle `UserProfileFormField`.

While this solution improves flexibility, it also complicates the user experience.

For Angular integration in [the upcoming release](https://github.com/keycloakify/keycloakify-angular), we've implemented this approach as it’s necessary; Angular doesn’t support multiple components per file like React does.   

But the whole aproach of ejecting pages is flawed.  

In practice, I've seen people ejecting everything, then struggling to update to newer versions because they’re now managing a lot of code they haven’t customized much and don’t fully understand. This is exactly what the "eject-page" command was meant to avoid—so I see this as a shortfall.

To tackle this differently, I’m experimenting with a new approach. The idea is a `npx keycloakify init` command that would automatically set up Keycloakify in your repo. Everything would be ejected by default, making it easier to understand what's going on. However, all files would be gitignored initially.

Keycloakify tooling would detect any manual changes to files and automatically remove them from `.gitignore`. This way, when updating to a new version, you’d get updated files for untouched pages, while needing to review only the files you’ve modified. We could even include command-line hints to guide users through the update process.

With this approach, we could significantly refactor, introducing a `src/login/components` directory with inputs and reusable components. This would make customization much simpler and enable us to provide alternative baselines, like `@keycloakify/paternfly` (the current one), `@keycloakify/mui`, `@keycloakify/shadcn`, or others. During initialization, users could choose the stack they prefer.

Realizing this vision will take time, so for now, it’s best to maintain the current structure to avoid technical debt or breaking changes.

**Comment by nima70** [comment]
Hey @garronej ,

Thanks for the detailed message! I found your problem really interesting—and, honestly, kind of a nightmare scenario!

Please go easy on me with this suggestion—I’m not as experienced on the front-end side, and I’m not really an open-source person. But have you thought about using a third-party dependency injection library, like **Inversify**?

The idea is to preinitialize the container with default dependencies so users can modify or override them if they want, then inject the container itself through `KcContext` to all pages and components. We could even expand your scripts to generate components based on this setup. I’ve used similar patterns on the backend and even in some smaller-scale React projects with good results, so I thought it might be worth mentioning.

As you mentioned earlier:

> "In practice, I've seen people ejecting everything, then struggling to update to newer versions because they’re now managing a lot of code they haven’t customized much and don’t fully understand. This is exactly what the 'eject-page' command was meant to avoid—so I see this as a shortfall."

Let’s look at a case study to show why I believe my suggestion could help: Right now, if I want to change something like `PasswordWrapper` on the register page, I’d need to create a new register page, wire it up in `KcPage`, modify `PasswordWrapper`, and then realize that the changes haven’t been applied everywhere. So, to change a single component, we end up having to refactor anything connected to that dependency. And that’s just the start! It becomes even more challenging if it’s passed to others who use it for years, as it becomes almost impossible to phase out. When users with various backgrounds are making changes, it’s even harder.

However, if we use some form of dependency injection like Inversify, we could switch a small part of the code seamlessly, even enabling easy transitions between different styles—like version 2 and legacy version 1.

Of course, as you mentioned:

> "While this solution improves flexibility, it also complicates the user experience."

I really love this statement. Not everyone understands dependency injection, and it has its own learning curve, which might not be ideal for an open-source project. But I’m not very experienced in open source, so I trust that you’d have a better sense of whether this makes sense or not.

Regarding this:

> "Keycloakify tooling would detect any manual changes to files and automatically remove them from .gitignore. This way, when updating to a new version, you’d get updated files for untouched pages, while needing to review only the files you’ve modified. We could even include command-line hints to guide users through the update process."

To be honest, I’m not entirely sure about this approach.

I also love this idea ❤️❤️❤️❤️:

> "With this approach, we could significantly refactor, introducing a `src/login/components` directory with inputs and reusable components. This would make customization much simpler and enable us to provide alternative baselines, like `@keycloakify/patternfly` (the current one), `@keycloakify/mui`, `@keycloakify/shadcn`, or others. During initialization, users could choose the stack they prefer."

If you decide to go this route, I’d be more than happy to help with the ShadCN setup!


And at the end, I 100% agree with you. Sure, let's stick to the current structure for now. Lool...

After all, I personally don’t have the time or resources to help with such a huge modification. So, I fully agree with your point!

loool… haha… 😆

**Comment by garronej** [comment]
I've spent considerable time thinking about dependency injection and exploring ways to make it work, but I've concluded it's a dead end for several reasons:

- **Userbase skill level**: Most of our users are junior developers or interns assigned the "easy task" of customizing Keycloak login pages. They often have limited experience and only a basic understanding of web programming. Adding more abstractions increases the barrier to entry for these users, making the project feel inaccessible. I want to ensure that understanding dependency injection isn't a prerequisite for creating a Keycloak theme.

- **Performance concerns**: Every byte counts in our use case. The page payload must be as light as possible to ensure fast load times, which are crucial for retention—even more so than style. While some problems are easy to solve on the backend, they become challenging on the frontend. We need to lazily load only the necessary components, which is tricky to implement with dependency injection, especially when trying to integrate React Suspense.

- **Type extension complexity**: In Keycloakify, users can add custom message keys, `kcContext` properties, and pages. These type extensions need to propagate to the components, requiring not only runtime dependency injection but also type-level dependency injection. This quickly becomes [nightmarish](https://github.com/keycloakify/keycloakify/blob/main/src/login/i18n/withJsx/i18nBuilder.ts).

> Honestly, I’m still not entirely sure about this approach.

While I understand this might appear as a less elegant solution, I believe it will make things easier for users. They’ll be able to view and modify code directly, while still benefiting from smooth updates. I don’t mind having code I didn’t write in my codebase as long as it’s not tracked by Git. In this case, we could dynamically edit a nested `.gitignore` (not the root one) to avoid tracking unnecessary files.

This approach would address the three issues I've been struggling with for years and eliminate those challenges entirely.

Another key advantage is that it would enable us to decouple Keycloakify as a build tool from Keycloakify as a component library. This would allow for an ecosystem of base Keycloak themes to be created independently from the core project. My vision includes supporting not only multiple component libraries and stacks but also multiple frameworks. 

For example, creating `@keycloakify/angular` took a lot of effort. Converting a project like [yours](https://github.com/nima70/keycloakify-tailwind-shadcn) into `@keycloakify/shadcn` would require significant work in the current paradigm. With the `.gitignore` approach, it could be as simple as packaging everything into an NPM module.

Just to clarify, I’m not expecting you to get involved in this—it’s my responsibility. I just wanted to share my thoughts with you. :)

**Reply by nima70** [reply]
Thank you for your comments. I found them very valuable.

I 100% agree with the first point.

On the second point, I agree with the performance trade-offs of my suggested methods. However, I’m not experienced enough to say whether the trade-offs are significant enough to be a complete dealbreaker. This is why I trust your expertise here—you’re certainly the better subject matter expert, and you’re most likely right about this, too.

Regarding the third point, I agree that the example you provided is tricky to handle. I think there might be room for improvement, but I don’t want to waste your time. Given your experience, I think your first and second points are strong enough to make my suggestion unnecessary.

BTW, my project isn’t great, to be honest. Knowing what I now understand about Keycloakify, I would never have written it this way. I initially started this project for myself, then decided to change course. A refactor is definitely needed.

Also, I really like your solution! If you need a hand, I’d be glad to help out. I’m extremely busy this month, so there’s been a bit of a delay, but I’ll fix the other issues soon.

To be honest, I’m learning a lot from you, and I really appreciate it. However, I don’t want to be a burden—lool. Writing comments like this is quite time-consuming, haha…

Anyway, thank you again, and have a wonderful evening!

**Reply by garronej** [reply]
Thanks for the feedback man!  

Don't wory, I've came very use to writing thoses kind of messaging.  
It helps me clarify my idea and help with the transparency of the project direction.  

> Also, I really like your solution! If you need a hand, I’d be glad to help out. I’m extremely busy this month, so there’s been a bit of a delay, but I’ll fix the other issues soon.  

You're help is welcome but don't stress about it! If you have time to fix the bugs you spotted the other day that would be great!

**Comment by nima70** [comment]
Thank you, @garronej, and Sure. 🙏 🙏 🙏

**Comment by uchar** [comment]
> Let’s look at a case study to show why I believe my suggestion could help: Right now, if I want to change something like PasswordWrapper on the register page, I’d need to create a new register page, wire it up in KcPage, modify PasswordWrapper, and then realize that the changes haven’t been applied everywhere. So, to change a single component, we end up having to refactor anything connected to that dependency. And that’s just the start! It becomes even more challenging if it’s passed to others who use it for years, as it becomes almost impossible to phase out. When users with various backgrounds are making changes, it’s even harder.

Yes in my case I want to convert it to Material ui and for exact same reasons as you said  I write a lot of codes that does most of this conversion automatically https://github.com/RionizSoftware/RionizKeycloakTheme/blob/main/src/codegen/MuiConvertor.ts
I also make sure all styles are in sperate files so that if eject page output changed in future I can atleast use styles https://github.com/RionizSoftware/RionizKeycloakTheme/tree/main/src/login/styles/pages 
I also had to change keycloakify code itself so that I can do more manual steps automatically ( like e.g adding id to component or ejecting all page in single command ,refactoring password wrapper and some other stuffs,... )  https://github.com/uchar/keycloakify/commits/main/

But now after two months I came back to it for upgrading Keycloakify and have some conflicts in my fork which is very time consuming to resolve and I dont even remember what I was doing back then lol so I prefer this stuffs handles once in this project instead of me doing same thing over and over again in the future

**Reply by garronej** [reply]
>  Keycloakify and have some conflicts in my fork which is very time consuming to resolve and I dont even remember what I was doing back then lol so I prefer this stuffs handles once in this project instead of me doing same thing over and over again in the future  

Well, Keycloakify now handles delegation of the `eject-page` and `add-story`. See [this reply](https://github.com/keycloakify/keycloakify/pull/663#issuecomment-2395600924) you can levrage it to implement your own ejection mecanism if you want.

**Comment by uchar** [comment]
>Assuming we have factorized the pages into smaller components, the eject command implementation can't simply be a script that copies page sources from the Keycloakify module into the user’s codebase anymore. It will need to analyze which components are in use and copy those as well, recursively.
That alone is complex. However, it still doesn't fully resolve the issue, as non-ejected pages will continue relying on internal components. To address this, we would need a mechanism to pass component implementations to pages, similar to how we handle UserProfileFormField.
While this solution improves flexibility, it also complicates the user experience. 

I think it's the right choice , analyzing component is actually not that complex I actually write it before for the reason I explained in my previous comment https://github.com/RionizSoftware/RionizKeycloakTheme/blob/main/src/codegen/typescriptUtilities/transformers/muImportAdder.ts 
Adding components to pages is also doable in the same way https://github.com/uchar/keycloakify/blob/71bab1f223869917a1eab385bc676c4b70a7ed8d/src/bin/tools/addImportAndSwitchToKCPage.ts 
I know @garronej have problem with adding typescript as a dependency to project but I think we can do it with other libraries too, it's definitely doable with babel parser which is already in project

about user experience , I'm pretty sure it would be much better than current state

**Comment by uchar** [comment]
>To tackle this differently, I’m experimenting with a new approach. The idea is a npx keycloakify init command that would automatically set up Keycloakify in your repo. Everything would be ejected by default, making it easier to understand what's going on. However, all files would be gitignored initially.
Keycloakify tooling would detect any manual changes to files and automatically remove them from .gitignore. This way, when updating to a new version, you’d get updated files for untouched pages, while needing to review only the files you’ve modified. We could even include command-line hints to guide users through the update process.

I dont think this approach solve any of the issues , main problem here is that if you change sth in one of the pages in the keycloakify itself (e.g a fix or sth) then people need to eject it again and chahnge those files again .( e.g rename tags to Material ui , add styles , add redux ,... ) this make project maintenance very hard .
Also I think most people need atleast half of those pages to be ejected to have a functional ui with their own theme 
But if we have better component structure and more SOLID approache we can even have another CLI e.g keycloakify eject-uiComponents so that it only eject components used by pages

We can even have different components with different libraries and more out of the box styles

**Reply by nima70** [reply]
I am not sure what would be the best way to achieve this 
>> But if we have better component structure and more SOLID approache we can even have another CLI e.g keycloakify eject-uiComponents so that it only eject components used by pages

However, I like the idea.
